src/webutil/tableState.js

```diff
@@ -0,0 +1,185 @@
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+  /**
+    A function that overwrites the base data and reapplies current sort and filter functions.
+    Sets pageIndex to 0.
+   */
+  +setData: ($ReadOnlyArray<T>) => void,
+|};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(defaults?: {
+  initialRowsPerPage?: number,
+  initialSort?: {
+    // The starting order of the initial sort, represented by "asc" or "desc".
+    sortOrder: SortOrderTypes,
+    // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    sortName: string,
+    // A function that returns the property of the element that should be used in the sort comparator.
+    sortFn: PluckFn<T>,
+  },
+}): TableState<T> {
+  defaults = defaults || {};
+  const sortDefaults = defaults.initialSort || {};
+  const [originalArray, setOriginalArray] = useState([]);
+  const [filteredArray, setFilteredArray] = useState([]);
+  const [sortName, setSortName] = useState(sortDefaults.sortName || "");
+  const [sortOrder, setSortOrder] = useState(
+    sortDefaults.sortOrder || SortOrders.ASC
+  );
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPage, setRowsPerPageInternal] = useState(
+    defaults.initialRowsPerPage || 20
+  );
+  let _sortFn: PluckFn<T> | null = sortDefaults.sortFn || null;
+  const _filterFns: Map<string, null | ((T) => boolean)> = new Map();
+
+  const _sortAndFilter = (_array: Array<T>, _sortOrder: string) => {
+    if (_sortFn) {
+      sortBy(_array, _sortFn);
+      if (_sortOrder === SortOrders.DESC) _array.reverse();
+    }
+    setOriginalArray(_array);
+    _filter(_array);
+  };
+
+  const _filter = (_array: Array<T>) => {
+    _filterFns.forEach((fn) => {
+      if (fn) _array = _array.filter(fn);
+    });
```

Comment:
Would be better to use .reduce here. Also shouldn't mutate the params, better to initialize a new variable:
```suggestion
    let filteredArray = _array;
    _filterFns.forEach((fn) => {
      if (fn) filteredArray = filteredArray.filter(fn);
    });
```

Reply:
accepted